File: Javascript/src/01.AlgorithmicToolbox/Week5DynamicProgramming1/editDistance.js
/**
 * Problem 3 - Edit Distance
 * Calculate the minimum number of operations (insertions, deletions and substitutions)
 * needed to transform one string to another.
 * https://en.wikipedia.org/wiki/Edit_distance
 * @param {string} first first string
 * @param {string} second second string
 * @returns {number} edit distance
 */
export function editDistance(first, second) {
  const editDistances = new Array(first.length + 1)
    .fill(Infinity)
    .map(() => new Array(second.length + 1).fill(Infinity));

  // intialize outer row / col
  // these are the number of operations to convert empty string to each input string
  for (let row = 0; row <= first.length; ++row) {
    editDistances[row][0] = row;
  }
  for (let col = 0; col <= second.length; ++col) {
    editDistances[0][col] = col;
  }
  // fill rest of elements
  for (let row = 1; row <= first.length; ++row) {
    for (let col = 1; col <= second.length; ++col) {
      if (first.charAt(row - 1) === second.charAt(col - 1)) {
        // same character, do nothing
        editDistances[row][col] = editDistances[row - 1][col - 1];
      } else {
        // with first input as reference; transform first to second
        const substitution = editDistances[row - 1][col - 1] + 1;
        const deletion = editDistances[row][col - 1] + 1;
        const addition = editDistances[row - 1][col] + 1;
        editDistances[row][col] = Math.min(substitution, deletion, addition);
      }
    }
  }
  return editDistances[first.length][second.length];
}

File: Rust/src/_01_algorithmic_toolbox/week5_dynamic_programming1/edit_distance.rs
/// Problem 3 - Edit Distance
/// Calculate the minimum number of operations
/// (insertions, deletions and substitutions)
/// needed to transform one string to another.
/// Assumes each character is of byte length!
/// Return minimum edit distance.
pub fn edit_distance(first: &String, second: &String) -> usize {
    let mut edit_distances = vec![vec![usize::MAX; second.len() + 1]; first.len() + 1];

    // initialize outer row / col
    // these are the number of operations to convert empty string to each input string
    for row in 0..=first.len() {
        edit_distances[row][0] = row;
    }
    for col in 0..=second.len() {
        edit_distances[0][col] = col;
    }
    // fill rest of the elements
    for row in 1..=first.len() {
        for col in 1..=second.len() {
            if first.as_bytes()[row - 1] == second.as_bytes()[col - 1] {
                edit_distances[row][col] = edit_distances[row - 1][col - 1];
            } else {
                let substitution = edit_distances[row - 1][col - 1] + 1;
                let deletion = edit_distances[row][col - 1] + 1;
                let addition = edit_distances[row - 1][col] + 1;
                edit_distances[row][col] =
                    std::cmp::min(std::cmp::min(substitution, deletion), addition);
            }
        }
    }
    edit_distances[first.len()][second.len()]
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn returns_minimum_edit_distance() {
        assert_eq!(edit_distance(&String::from("ab"), &String::from("ab")), 0);
        assert_eq!(
            edit_distance(&String::from("short"), &String::from("ports")),
            3
        );
        assert_eq!(
            edit_distance(&String::from("editing"), &String::from("distance")),
            5
        );
        assert_eq!(
            edit_distance(&String::from("abcdef"), &String::from("azced")),
            3
        );
    }
}
